compiler/passes/parallel.cpp

```diff
@@ -255,14 +265,54 @@ static Symbol* insertAutoCopyDestroyForTaskArg
         insertReferenceTemps(autoDestroyCall);
       }
     }
-    else if (isRecord(baseType))
+    else
     {
-      // Do this only if the record is passed by value.
-      if (arg->typeInfo() == baseType)
+      INT_ASSERT(needForRecord);
+      // This gotta be a task function, possibly for an 'on'.
+      // Otherwise why are we here?
+      INT_ASSERT(isTaskFun(fn));
+
+      // TODO: Find out why _RuntimeTypeInfo records do not have autoCopy
+      // functions, so we can get rid of this special test.
+      if (autoCopyFn == NULL)
+      {
+        // nothing to do
+      }
+      else if (autoCopyFn->hasFlag(FLAG_COMPILER_GENERATED) &&
+               autoDestroyFn->hasFlag(FLAG_COMPILER_GENERATED))
+      {
+        // If so, autoCopy/autoDestroy have no side effects
+        // (see markPODtypes and FLAG_HAS_USER_DESTRUCTOR),
+        // so we do not need to insert them, and do not need
+        // the USR_FATAL_CONT in the "else" case.
+      }
+      else if (fn->hasFlag(FLAG_ON) && arg == fcall->get(1))
+      {
+        // Also we do not want to generate the USR_FATAL_CONT
+        // when passing dummy_locale_arg: chpl_localeID_t,
+        // which is a record that is passed to any on_fn.
+        //
+        // TODO: annotate chpl_localeID_t's autocopy/destroy
+        // with FLAG_COMPILER_GENERATED to avoid this special case?
+      }
+      else
       {
-        // TODO: Find out why _RuntimeTypeInfo records do not have autoCopy
-        // functions, so we can get rid of this special test.
-        if (autoCopyFn == NULL) return var;
+        // Beware of record destructors with 'on', e.g.:
+        //   proc ~R.R() { on ... { ... this ... } }
+        //
+        // There, currently 'this' will be autoCopy-ed into the on_fn,
+        // then autoDestroy-ed there. That will call the destructor,
+        // autoDestroy-ing, etc. -- resulting in an infinite loop.
+        //
+        // The bug is that we pass records by value, whereas it should be
+        // by const ref - in which case there will be no construction or
+        // destruction, so no infinite loop.
+        //
```

Comment:
Yeah.  That's unfortunate.
Blank intent should be interpreted as "const ref" for records, so there are just a few special (or compiler-inserted) cases that need to be made consistent with that.